### DISCUSSION TITLE
Extract images from docx documents

### DISCUSSION THREAD

**Discussion by mmalkus** [davidgohel/officer#450] [Ideas]
Thanks for the great work on this package.
I noticed that for powerpoint you can easily extract images. For word not.
Looking into the actual xml in a docx file, it seems to be reasonably simple to do.
Not knowing how simple this would be to implement I'd say you could do the following:

- identify images in document.xml (rId XX)
- find the corresponding path in document.xml.rels
- insert new rows in the docx_summary output

**Comment by davidgohel** [comment]
yes, it's true :)

If you want to PR something, it will be welcome. That is not that complicated, you would need a bit of diving into the paragraph and table Word XML structure, but it seems you already got that. 

Actually, it returns contents at 'block' level (tables and paragraph). An option should be provided to provide a more detailed content, at the 'chunk' level (text and images within paragraphs). That way also, nothing existing will break because of this change.

If you don't want to propose a PR I can set that as an issue with label 'improvments' but it may take time to be implemented as I am not focused on that at the moment.

**Comment by mmalkus** [comment]
Not sure how to do that exactly, will give it a try.

There seems to be already some stuff implemented, but not working. Specifically on line 68 of https://github.com/davidgohel/officer/blob/master/R/relationship.R there appears to be an error
is_img <- basename( private$type ) %in% "image"
should be 
is_img <- basename( private$type ) %in% "http://schemas.openxmlformats.org/officeDocument/2006/relationships/image"

Not sure if it magically starts working then though

**Reply by davidgohel** [reply]
`basename("http://schemas.openxmlformats.org/officeDocument/2006/relationships/image")` is returning "image". If you doubt the correct implementation of some bit of codes, you can contribute by proposing unit tests.

**Reply by mmalkus** [reply]
Thanks for the correction :) I'll try to make it work first

**Reply by davidgohel** [reply]
As I said, prove there is an error before correcting it :) Until now, I don't see any evidence that the PowerPoint function is not able to correctly identify image path

**Reply by mmalkus** [reply]
I essentially now got code that does what want it to be doing (and works), however this clearly cannot directly be implemented in your code. Can you maybe give me some directions on what functions I should be modifying?

library(xml2)

mapping <- sample_data$doc_obj$rel_df()
mapping$type <- basename(mapping$type)
mapping <- mapping %>% tibble %>% filter(type == "image")

all_nodes <- xml_find_all(sample_data$doc_obj$get(), "/w:document/w:body/*[self::w:p or self::w:tbl]")
image_nodes <- xml_find_all(all_nodes, "//a:blip[@r:embed]")

image_details <- tibble(id = xml_attr(image_nodes, "embed")) %>% left_join(mapping, by = "id")

a <- c()
for(i in seq_along(all_nodes)) {
if (length(xml_find_all(all_nodes[i], ".//a:blip[@r:embed]") )> 0)  a <- c(a,i)
}

content <- docx_summary(sample_data)
content$style_name[content$doc_index %in% a] <- "figure"
content$text[content$doc_index %in% a] <- image_details$target
